File: C++/8-Graphs(BFS-DFS)/Algorithms/3-Dijkstra’sAlgorithm(ShortestPath)/Dijkstra’sAlgorithm(ShortestPath).cpp
#include <iostream>
#include <vector>
#include <queue>
#include <climits>
#include <functional>
#include <algorithm>
#include <set>
#include <map>

using namespace std;

// Structure to represent an edge in the graph
struct Edge {
    int to;      // Destination vertex
    int weight;    // Weight of the edge
};

// Structure to represent a node in the priority queue for Dijkstra's algorithm
struct Node {
    int vertex;  // Vertex
    int distance; // Distance from the source

    // Overload the less than operator for the priority queue.  We want the *smallest* distance
    // to have the highest priority, so we reverse the comparison.
    bool operator>(const Node& other) const {
        return distance > other.distance;
    }
};

// Function to add an edge to the graph.  We represent the graph as an adjacency list.
void addEdge(vector<vector<Edge>>& graph, int from, int to, int weight) {
    graph[from].push_back({to, weight});
    // For an undirected graph, you would also add the reverse edge:
    // graph[to].push_back({from, weight});
}

// 1. Dijkstra's Algorithm using a Priority Queue (Min Heap)
//    - Uses priority_queue (which is a max heap by default in C++)
//    - We use std::greater to make it a min heap.
//    - Efficient for most cases.
vector<int> dijkstraPQ(const vector<vector<Edge>>& graph, int source) {
    int numVertices = graph.size();
    vector<int> distance(numVertices, INT_MAX); // Initialize distances to infinity
    distance[source] = 0;                      // Distance from source to itself is 0

    priority_queue<Node, vector<Node>, greater<Node>> pq; // Min heap
    pq.push({source, 0});

    while (!pq.empty()) {
        Node current = pq.top();
        pq.pop();
        int u = current.vertex;
        int dist_u = current.distance;

        if (dist_u > distance[u]) {
            continue; // Important optimization: Skip if we've found a shorter path to u
        }

        for (const Edge& edge : graph[u]) {
            int v = edge.to;
            int weight_uv = edge.weight;
            if (distance[u] != INT_MAX && distance[u] + weight_uv < distance[v]) {
                distance[v] = distance[u] + weight_uv; // Update distance
                pq.push({v, distance[v]});           // Add to priority queue
            }
        }
    }
    return distance;
}

// 2. Dijkstra's Algorithm using a Set (for ordered storage)
//    - Uses a set to store nodes and their distances.
//    - Set automatically orders elements, providing the smallest distance at the beginning.
vector<int> dijkstraSet(const vector<vector<Edge>>& graph, int source) {
    int numVertices = graph.size();
    vector<int> distance(numVertices, INT_MAX);
    distance[source] = 0;

    set<pair<int, int>> s; // Store {distance, vertex} pairs.  Set is ordered by the first element.
    s.insert({0, source});

    while (!s.empty()) {
        pair<int, int> current = *s.begin(); // Get the node with the smallest distance
        s.erase(s.begin());             // Remove it from the set
        int u = current.second;
        int dist_u = current.first;

       if (dist_u > distance[u]) {
            continue; // Important optimization: Skip if we've found a shorter path to u
        }

        for (const Edge& edge : graph[u]) {
            int v = edge.to;
            int weight_uv = edge.weight;
            if (distance[u] != INT_MAX && distance[u] + weight_uv < distance[v]) {
                // If a shorter path is found, we need to update the distance.
                // Because set does not allow updating the element directly, we remove the old pair and insert a new one.
                s.erase({distance[v], v}); // Remove the old distance-vertex pair
                distance[v] = distance[u] + weight_uv;
                s.insert({distance[v], v}); // Insert the updated pair
            }
        }
    }
    return distance;
}

// 3. Dijkstra's Algorithm using a simple array (less efficient, but good for understanding)
//    -  The most basic implementation.
//    -  Finds the vertex with the minimum distance by iterating through all vertices.
//    -  Time complexity: O(V^2), where V is the number of vertices.
vector<int> dijkstraArray(const vector<vector<Edge>>& graph, int source) {
    int numVertices = graph.size();
    vector<int> distance(numVertices, INT_MAX);
    distance[source] = 0;
    vector<bool> visited(numVertices, false); // Keep track of visited vertices

    for (int count = 0; count < numVertices - 1; ++count) {
        // Find the vertex with the minimum distance among the unvisited vertices
        int minDistance = INT_MAX, u;
        for (int v = 0; v < numVertices; ++v) {
            if (!visited[v] && distance[v] <= minDistance) {
                minDistance = distance[v];
                u = v;
            }
        }

        visited[u] = true; // Mark the selected vertex as visited

        // Update distances of the adjacent vertices of u
        for (const Edge& edge : graph[u]) {
            int v = edge.to;
            int weight_uv = edge.weight;
            if (distance[u] != INT_MAX && distance[u] + weight_uv < distance[v]) {
                distance[v] = distance[u] + weight_uv;
            }
        }
    }
    return distance;
}

// 4. Dijkstra with Adjacency Matrix
//    - Uses an adjacency matrix to represent the graph.  Useful for dense graphs.
//    -  Adjacency Matrix Graph Representation:
//       graph[u][v] stores the weight of the edge from vertex u to vertex v.
//       If there is no edge between u and v, graph[u][v] is 0 or a large value (like INT_MAX).
vector<int> dijkstraMatrix(const vector<vector<pair<int, int>>>& graph, int source) {
    int numVertices = graph.size();
    vector<int> distance(numVertices, INT_MAX);
    distance[source] = 0;
    vector<bool> visited(numVertices, false);

    for (int count = 0; count < numVertices - 1; ++count) {
        int minDistance = INT_MAX, u;
        for (int v = 0; v < numVertices; ++v) {
            if (!visited[v] && distance[v] <= minDistance) {
                minDistance = distance[v];
                u = v;
            }
        }
        visited[u] = true;

        for (int v = 0; v < numVertices; ++v) {
            // Important:  Check if there is an edge from u to v AND handle the case where there is no edge.
            if (graph[u][v].first != -1 && distance[u] != INT_MAX && distance[u] + graph[u][v].second < distance[v]) {
                distance[v] = distance[u] + graph[u][v].second;
            }
        }
    }
    return distance;
}

// 5. Dijkstra's Algorithm using a Fibonacci Heap (Conceptual)
//    -  Conceptually uses a Fibonacci heap.  C++ doesn't have a standard Fibonacci heap.
//    -  Fibonacci heaps offer better theoretical time complexity, especially for sparse graphs.
//    -  In practice, they are more complex to implement and the constant factors can make them slower for small to medium-sized graphs.
//    -  This implementation uses a map to simulate the decrease-key operation (which is the key advantage of Fibonacci heaps).
//    -  This is more of a conceptual implementation, as it doesn't use a true Fibonacci Heap.
vector<int> dijkstraFibonacci(const vector<vector<Edge>>& graph, int source) {
    int numVertices = graph.size();
    vector<int> distance(numVertices, INT_MAX);
    distance[source] = 0;

    map<int, int> m; // Map to store {distance, vertex}.  Ordered by distance.
    m[0] = source;

    while (!m.empty()) {
        auto it = m.begin();
        int u = it->second;
        int dist_u = it->first;
        m.erase(it);

        if (dist_u > distance[u]) {
            continue;
        }

        for (const Edge& edge : graph[u]) {
            int v = edge.to;
            int weight_uv = edge.weight;
            if (distance[u] != INT_MAX && distance[u] + weight_uv < distance[v]) {
                // "Decrease-key" operation:  We don't have a true decrease-key in std::map, so we remove and re-insert.
                if (m.find(distance[v]) != m.end()) {
                    m.erase(distance[v]);
                }
                distance[v] = distance[u] + weight_uv;
                m[distance[v]] = v; // Insert with the new, smaller distance
            }
        }
    }
    return distance;
}

int main() {
    int numVertices = 6;
    vector<vector<Edge>> graph(numVertices);

    // Build the graph (example graph)
    addEdge(graph, 0, 1, 10);
    addEdge(graph, 0, 2, 5);
    addEdge(graph, 1, 2, 2);
    addEdge(graph, 1, 3, 1);
    addEdge(graph, 2, 1, 3);
    addEdge(graph, 2, 3, 9);
    addEdge(graph, 2, 4, 2);
    addEdge(graph, 3, 4, 4);
    addEdge(graph, 3, 5, 6);
    addEdge(graph, 4, 5, 4);

    // Example of Adjacency Matrix representation
    vector<vector<pair<int, int>>> matrixGraph(numVertices, vector<pair<int, int>>(numVertices, {-1, 0})); // Initialize with -1 and 0
    // Convert the previous graph to adjacency matrix
    for (int i = 0; i < numVertices; i++) {
        for (const Edge& edge : graph[i]) {
            matrixGraph[i][edge.to] = {1, edge.weight}; // 1 indicates there is an edge, second value is the weight.
        }
    }
    // Choose the source vertex
    int source = 0;

    // 1. Dijkstra with Priority Queue
    vector<int> distancePQ = dijkstraPQ(graph, source);
    cout << "Dijkstra's Algorithm with Priority Queue (Min Heap):" << endl;
    for (int i = 0; i < numVertices; ++i) {
        cout << "Vertex " << i << ": Distance = " << distancePQ[i] << endl;
    }
    cout << endl;

    // 2. Dijkstra with Set
    vector<int> distanceSet = dijkstraSet(graph, source);
    cout << "Dijkstra's Algorithm with Set:" << endl;
    for (int i = 0; i < numVertices; ++i) {
        cout << "Vertex " << i << ": Distance = " << distanceSet[i] << endl;
    }
    cout << endl;

    // 3. Dijkstra with Array
    vector<int> distanceArray = dijkstraArray(graph, source);
    cout << "Dijkstra's Algorithm with Array:" << endl;
    for (int i = 0; i < numVertices; ++i) {
        cout << "Vertex " << i << ": Distance = " << distanceArray[i] << endl;
    }
    cout << endl;

    // 4. Dijkstra with Adjacency Matrix
    vector<int> distanceMatrix = dijkstraMatrix(matrixGraph, source);
    cout << "Dijkstra's Algorithm with Adjacency Matrix:" << endl;
    for (int i = 0; i < numVertices; ++i) {
        cout << "Vertex " << i << ": Distance = " << distanceMatrix[i] << endl;
    }
    cout << endl;

    // 5. Dijkstra with Fibonacci Heap (Conceptual)
    vector<int> distanceFibonacci = dijkstraFibonacci(graph, source);
    cout << "Dijkstra's Algorithm with Fibonacci Heap (Conceptual):" << endl;
    for (int i = 0; i < numVertices; ++i) {
        cout << "Vertex " << i << ": Distance = " << distanceFibonacci[i] << endl;
    }
    cout << endl;

    return 0;
}

File: Java/8-Graphs(BFS-DFS)/Algorithms/3-Dijkstra’sAlgorithm(ShortestPath)/Dijkstra’sAlgorithm(ShortestPath).java
import java.util.*;

public class DijkstraAlgorithm {

    // 1. Dijkstra's Algorithm using Priority Queue (Optimized)
    //    - Uses a PriorityQueue to efficiently select the node with the smallest distance.
    //    - Time Complexity: O(E + V log V), where V is the number of vertices and E is the number of edges.
    //    - This is the most common and efficient implementation.
    public static int[] dijkstraPQ(int[][] graph, int start) {
        int numVertices = graph.length;
        int[] distances = new int[numVertices];
        Arrays.fill(distances, Integer.MAX_VALUE);
        distances[start] = 0;

        // PriorityQueue to store nodes with their distances.  The queue is ordered
        // based on the distance to the node (shortest distance first).
        PriorityQueue<Node> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a.distance));
        pq.offer(new Node(start, 0));

        while (!pq.isEmpty()) {
            Node u = pq.poll();
            int uVertex = u.vertex;
            int uDistance = u.distance;

            // Important optimization: If the current distance to u is greater than the
            // known shortest distance, we can skip processing its neighbors.  This is
            // because we've already found a shorter path to u.
            if (uDistance > distances[uVertex]) {
                continue;
            }

            for (int v = 0; v < numVertices; v++) {
                if (graph[uVertex][v] != 0) { // There is an edge from u to v
                    int edgeWeight = graph[uVertex][v];
                    int newDistance = uDistance + edgeWeight;

                    if (newDistance < distances[v]) {
                        distances[v] = newDistance;
                        pq.offer(new Node(v, newDistance));
                    }
                }
            }
        }
        return distances;
    }

    // 2. Dijkstra's Algorithm using a basic array (Unoptimized)
    //    - Uses a simple array to store distances and iterates through it to find
    //      the minimum distance.
    //    - Time Complexity: O(V^2), where V is the number of vertices.
    //    - Less efficient than the PriorityQueue approach, especially for large graphs.
    public static int[] dijkstraArray(int[][] graph, int start) {
        int numVertices = graph.length;
        int[] distances = new int[numVertices];
        boolean[] visited = new boolean[numVertices];
        Arrays.fill(distances, Integer.MAX_VALUE);
        distances[start] = 0;

        for (int count = 0; count < numVertices - 1; count++) {
            // Find the vertex with the minimum distance among the unvisited vertices
            int minDistance = Integer.MAX_VALUE;
            int u = -1; // Initialize u to -1 (invalid vertex)
            for (int v = 0; v < numVertices; v++) {
                if (!visited[v] && distances[v] <= minDistance) {
                    minDistance = distances[v];
                    u = v;
                }
            }

            // If no unvisited vertex with finite distance was found, exit the loop.
            if (u == -1) break;

            visited[u] = true; // Mark the selected vertex as visited

            // Update distances of the adjacent vertices of u
            for (int v = 0; v < numVertices; v++) {
                if (!visited[v] && graph[u][v] != 0 && distances[u] != Integer.MAX_VALUE
                        && distances[u] + graph[u][v] < distances[v]) {
                    distances[v] = distances[u] + graph[u][v];
                }
            }
        }
        return distances;
    }

    // 3. Dijkstra's Algorithm with Path Reconstruction
    //    -  In addition to finding the shortest distances, it also reconstructs the
    //       shortest paths.
    //    - Time Complexity: O(E + V log V)  (using PriorityQueue)
    public static Map<Integer, Integer> dijkstraWithPath(int[][] graph, int start, int end) {
        int numVertices = graph.length;
        int[] distances = new int[numVertices];
        int[] predecessors = new int[numVertices]; // Store the predecessor node in the shortest path
        Arrays.fill(distances, Integer.MAX_VALUE);
        distances[start] = 0;
        Arrays.fill(predecessors, -1); // Initialize predecessors array

        PriorityQueue<Node> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a.distance));
        pq.offer(new Node(start, 0));

        while (!pq.isEmpty()) {
            Node u = pq.poll();
            int uVertex = u.vertex;
            int uDistance = u.distance;

            if (uDistance > distances[uVertex]) {
                continue;
            }

            if (uVertex == end) {
                break; // Optimization: Stop if we reach the end node.
            }

            for (int v = 0; v < numVertices; v++) {
                if (graph[uVertex][v] != 0) {
                    int edgeWeight = graph[uVertex][v];
                    int newDistance = uDistance + edgeWeight;

                    if (newDistance < distances[v]) {
                        distances[v] = newDistance;
                        predecessors[v] = uVertex; // Store the predecessor
                        pq.offer(new Node(v, newDistance));
                    }
                }
            }
        }

        // Reconstruct the path from start to end using the predecessors array
        Map<Integer, Integer> path = new HashMap<>();
        if (distances[end] != Integer.MAX_VALUE) {
            int current = end;
            while (current != start) {
                path.put(predecessors[current], current);
                current = predecessors[current];
            }
        }
        return path;
    }

    // 4. Dijkstra with Adjacency List
    //    - Uses an adjacency list to represent the graph, which can be more efficient
    //      for sparse graphs (graphs with fewer edges).
    //    - Time Complexity: O(E + V log V)
    public static int[] dijkstraAdjacencyList(List<List<Edge>> adjList, int start) {
        int numVertices = adjList.size();
        int[] distances = new int[numVertices];
        Arrays.fill(distances, Integer.MAX_VALUE);
        distances[start] = 0;

        PriorityQueue<Node> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a.distance));
        pq.offer(new Node(start, 0));

        while (!pq.isEmpty()) {
            Node u = pq.poll();
            int uVertex = u.vertex;
            int uDistance = u.distance;

            if (uDistance > distances[uVertex]) {
                continue;
            }

            for (Edge edge : adjList.get(uVertex)) {
                int v = edge.destination;
                int weight = edge.weight;
                int newDistance = uDistance + weight;

                if (newDistance < distances[v]) {
                    distances[v] = newDistance;
                    pq.offer(new Node(v, newDistance));
                }
            }
        }
        return distances;
    }

    // 5. Dijkstra's Algorithm for a Grid (2D array)
    //    -  Handles a 2D grid where each cell represents a node, and movement
    //       is allowed in four directions (up, down, left, right).
    //    - Time Complexity: O(R * C * log(R * C)) where R is number of rows and C is number of columns.
    public static int dijkstraGrid(int[][] grid, int startRow, int startCol, int endRow, int endCol) {
        int rows = grid.length;
        int cols = grid[0].length;
        int[][] distances = new int[rows][cols];
        for (int[] row : distances) {
            Arrays.fill(row, Integer.MAX_VALUE);
        }
        distances[startRow][startCol] = 0;

        PriorityQueue<GridNode> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a.distance));
        pq.offer(new GridNode(startRow, startCol, 0));

        int[] dr = {-1, 1, 0, 0}; // Possible row movements (up, down, left, right)
        int[] dc = {0, 0, -1, 1}; // Possible column movements

        while (!pq.isEmpty()) {
            GridNode u = pq.poll();
            int uRow = u.row;
            int uCol = u.col;
            int uDistance = u.distance;

            if (uDistance > distances[uRow][uCol]) {
                continue;
            }

            if (uRow == endRow && uCol == endCol) {
                return uDistance; // Found the shortest path
            }

            for (int i = 0; i < 4; i++) {
                int newRow = uRow + dr[i];
                int newCol = uCol + dc[i];

                if (newRow >= 0 && newRow < rows && newCol >= 0 && newCol < cols) {
                    int newDistance = uDistance + grid[newRow][newCol]; // Assuming grid[r][c] is the cost to move to cell (r, c)
                    if (newDistance < distances[newRow][newCol]) {
                        distances[newRow][newCol] = newDistance;
                        pq.offer(new GridNode(newRow, newCol, newDistance));
                    }
                }
            }
        }
        return -1; // No path found
    }

    // Helper class for Dijkstra with Priority Queue and Adjacency List
    static class Node {
        int vertex;
        int distance;

        public Node(int vertex, int distance) {
            this.vertex = vertex;
            this.distance = distance;
        }
    }

    // Helper class for Dijkstra in Grid
    static class GridNode {
        int row;
        int col;
        int distance;

        public GridNode(int row, int col, int distance) {
            this.row = row;
            this.col = col;
            this.distance = distance;
        }
    }

    // Helper class for Adjacency List
    static class Edge {
        int destination;
        int weight;

        public Edge(int destination, int weight) {
            this.destination = destination;
            this.weight = weight;
        }
    }

    public static void main(String[] args) {
        // Example graph represented as an adjacency matrix
        int[][] graph = {
                {0, 4, 0, 0, 0, 0, 0, 8, 0},
                {4, 0, 8, 0, 0, 0, 0, 11, 0},
                {0, 8, 0, 7, 0, 4, 0, 0, 2},
                {0, 0, 7, 0, 9, 14, 0, 0, 0},
                {0, 0, 0, 9, 0, 10, 0, 0, 0},
                {0, 0, 4, 14, 10, 0, 2, 0, 0},
                {0, 0, 0, 0, 0, 2, 0, 1, 6},
                {8, 11, 0, 0, 0, 0, 1, 0, 7},
                {0, 0, 2, 0, 0, 0, 6, 7, 0}
        };

        int startNode = 0;
        System.out.println("Dijkstra's Algorithm Results:");

        // 1. Dijkstra with Priority Queue
        int[] distancesPQ = dijkstraPQ(graph, startNode);
        System.out.println("Shortest distances from node " + startNode + " (Priority Queue): " + Arrays.toString(distancesPQ));

        // 2. Dijkstra with Array
        int[] distancesArray = dijkstraArray(graph, startNode);
        System.out.println("Shortest distances from node " + startNode + " (Array): " + Arrays.toString(distancesArray));

        // 3. Dijkstra with Path Reconstruction
        int endNode = 6;
        Map<Integer, Integer> path = dijkstraWithPath(graph, startNode, endNode);
        System.out.println("Shortest path from " + startNode + " to " + endNode + ": " + getPathString(path, startNode, endNode));

        // 4. Dijkstra with Adjacency List
        List<List<Edge>> adjList = new ArrayList<>();
        for (int i = 0; i < graph.length; i++) {
            adjList.add(new ArrayList<>());
            for (int j = 0; j < graph[i].length; j++) {
                if (graph[i][j] != 0) {
                    adjList.get(i).add(new Edge(j, graph[i][j]));
                }
            }
        }
        int[] distancesList = dijkstraAdjacencyList(adjList, startNode);
        System.out.println("Shortest distances from node " + startNode + " (Adjacency List): " + Arrays.toString(distancesList));

        // 5. Dijkstra for Grid
        int[][] grid = {
                {0, 3, 0, 0},
                {1, 0, 2, 3},
                {0, 2, 0, 1},
                {0, 0, 1, 0}
        };
        int startRow = 0, startCol = 0;
        int endRow = 3, endCol = 3;
        int shortestDistanceGrid = dijkstraGrid(grid, startRow, startCol, endRow, endCol);
        System.out.println("Shortest distance in grid from (" + startRow + "," + startCol + ") to (" + endRow + "," + endCol + "): " + shortestDistanceGrid);
    }

    // Helper method to reconstruct path as a string
    private static String getPathString(Map<Integer, Integer> path, int start, int end) {
        if (path.isEmpty()) {
            return "No path exists";
        }
        StringBuilder sb = new StringBuilder();
        sb.append(start);
        int current = start;
        while (current != end) {
            current = path.get(current);
            sb.append(" -> ").append(current);
        }
        return sb.toString();
    }
}

